### PR TITLE
Enables multinode calibration flow through calibrate_model.sh

### PR DESCRIPTION
**Pull request by vishnumadhu365**
This PR enables support for multinode calibration using Ray as the distributed executor backend.
contributors : @parthabas 

Tested on : 

- 2-node Gaudi2 for FP8 quantized serving of meta-llama/Llama-3.1-405B-Instruct
- validated on all 1.19.1 stack
- vllm-fork (habana_main - 397ec534abb271574a4b22e2a78fb80c93291f03)
- vault.habana.ai/gaudi-docker/1.19.1/ubuntu22.04/habanalabs/pytorch-installer-2.5.1:latest

Updates : 

- included section for multinode training in calibration/README.md
- update measure_scales.py, quantize_scales.py to expose vllm distributed executor backend
- update calibrate_model.sh to include the multinode checks and execution
- fix unify_measurement script to address missing measurement files, wrong pattern matching etc.

Open issue : 
- unification script fails for unify to 1X cards (HS-4988) - Will be addressed in a diff PR

### CODE REVIEW

File: calibration/calibrate_model.sh [hunk 1]
```diff
@@ -130,8 +169,20 @@ python step-3-postprocess_measure.py -m $FP8_DIR/$MODEL_NAME/$DEVICE_TYPE/ -o in
...
+    sleep 2
+else
+    export QUANT_CONFIG=$FP8_DIR/$MODEL_NAME/maxabs_quant_$DEVICE_TYPE.json
+fi
+
 echo ""
 echo "4/4 Quantize scales"
-export QUANT_CONFIG=$FP8_DIR/$MODEL_NAME/maxabs_quant_$DEVICE_TYPE.json
-python step-4-quantize-scales.py --model $MODEL_PATH --tensor-parallel-size $TP_SIZE || (echo "Error in step 4" && exit 1)
+if $MULTI_NODE_RUN; then
```
Comment (nirda7) [inline calibration/calibrate_model.sh:181]: If it is a multi node run and we want to use unification then we shouldn't run the quantization here.
We want to have
Measurements-Unification-Quantization
and not:
Measurements-Quantization-Unification-Quantization

I think we can add here the unification process as part of the calibrate_model.sh script. (using another parameter/flag of -u and grouping - just making sure -u and -t values make sense together)

**Review by nirda7 — CHANGES_REQUESTED** [review]

File: calibration/calibrate_model.sh [hunk 2]
```diff
@@ -22,6 +22,10 @@ usage() {
     echo
 }
 
+cleanup_tmp() {
```
[the remark this answers is not in the document]
Comment (vishnumadhu365) [reply calibration/calibrate_model.sh:26]: changes made :
included safety check on pwd , 
calibrate_model.sh does not seem to be designed to be executed from outside the parent repo dir, the python invocations will fail otherwise. change script execution dir at the top of script

File: calibration/README.md [hunk 3]
```diff
@@ -30,3 +30,88 @@ Here are some examples of how to use the script:
...
+# Add worker nodes to the Ray cluster
+ray start --address='<ip-of-ray-head-node>:6379'
+
+# Check if the cluster has required number of HPU's
+ray status
+```
+
+#### Step 3: Run model calibration script
+```bash
+./calibrate_model.sh -m meta-llama/Llama-3.1-405B-Instruct -d <path-to-dataset>/open_orca_gpt4_tokenized_llama.calibration_1000.pkl -o <path-to-calibration-output>/fp8_output -l 10 -t 16 -b 1
```
[the remark this answers is not in the document]
Comment (vishnumadhu365) [reply calibration/README.md:74]: Just gave a sample command, users are free to experiment with batch_size and num_samples based on specific model and usecase. Should this be changed to match some defaults ?

Does step-4-quantize-scales.py persist any changes ? seems its only loading the base model >> apply measured scales with inc and shutting down, thus validating the quantized model loading works fine, pls correct if this understanding is wrong.

File: calibration/README.md [hunk 4]
```diff
@@ -30,3 +30,88 @@ Here are some examples of how to use the script:
...
+./manage_network_ifs.sh --down 
+./manage_network_ifs.sh --up 
+```  
+  </details>
+
+
+#### Step 4: (optional) Measurement unification <p>
+This is an optional step and is used to reduce the target tensor parallelism level by unifying the measurement scales.<br> For eg: You can perform FP8 calibration on the Llama 3.1 405B model on 2x Gaudi2 nodes with Tensor Parallelism = 16 and then use the unification script to reduce the TP to 8. Refer sample command below
+```bash
+python step-5-unify_measurements.py -g "0,1--2,3--4,5--6,7--8,9--10,11--12,13--14,15"  -m <path-to-calibration-output>/fp8_output/llama-3.1-405b-instruct/g2/ -o ./unification_files_8x
```
[the remark this answers is not in the document]
Comment (vishnumadhu365) [reply calibration/README.md]: done

File: calibration/calibrate_model.sh (hunk 1 above)
Comment (vishnumadhu365) [reply calibration/calibrate_model.sh:187]: Need to understand the role of "Quantization" step in more detail. 
For 405B we have been following , Measurements(tp16)-Quantization(tp16)-Unification(tp8)-vllm serve (tp8) without drop in accuracy. 

including unifying in calibrate_model script, can we target this in some future PR ?

Comment (nirda7) [reply calibration/calibrate_model.sh:187]: Quantization step calculate the scales out of the measurements provided from steps 2 and 3.
It initializes the LLM object with the converted model (after being patched with the fp8 submodules) and then run all the warmup steps - so basically it's a sanity test for the quantized model on the vllm infrastructure. 

So I believe running the quantization on the same "environment" on which you are going to run your inference will make more sense (as being a sanity check) than running the quantization on 2 nodes and only during inference run it on 1 node...

did you compare results when using unification before quant and after?

about whether and when to add the unification process to the calibration process I'll suggest to consult with Artur/Oz/Michal K...

Comment (vishnumadhu365) [reply calibration/calibrate_model.sh:187]: unifications was always ran after quantization, haven't tested the before scenario

File: calibration/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (vishnumadhu365) [reply calibration/README.md]: updated with default values of args

File: calibration/calibrate_model.sh [hunk 5]
```diff
@@ -77,6 +90,27 @@ while getopts "m:b:l:t:d:h:o:" OPT; do
     esac
 done
 
+if [[ $TP_SIZE -gt 8 ]]; then
```
Comment (afierka-intel) [inline calibration/calibrate_model.sh:93]: Move this all multi-node checks after the line 114, i.e. after this check:
```
if [[ -z "$MODEL_PATH" && -z "$FP8_DIR" && -z "$DATASET_PATH" ]]; then
    echo "Model stub, source dataset path and output path for fp8 measurements must be provided."
    usage

    
        
          
    

        
        Expand All
    
    @@ -87,7 +121,7 @@ fi
  
    exit 1
fi
```

Multi-node is still optional. Let's check required parameter first.

Comment (vishnumadhu365) [reply calibration/calibrate_model.sh:99]: done

File: calibration/README.md [hunk 6]
```diff
@@ -30,3 +30,88 @@ Here are some examples of how to use the script:
 # Run inference with FP8 models
 
 An inference with FP8 precision models using vLLM has been described in [README_GAUDI](https://github.com/HabanaAI/vllm-fork/blob/habana_main/README_GAUDI.md#quantization-fp8-inference-and-model-calibration-process) file.
+
+# Multi-node FP8 Calibration 
+
+> [!WARNING] 
+> !! Mutli-node calibration is an experimental feature and could have stability issues.
```
Comment (afierka-intel) [inline calibration/README.md:37]: Are "!!" necessary there?

Comment (vishnumadhu365) [reply calibration/README.md]: not needed ,removed

File: calibration/calibrate_model.sh [hunk 7]
```diff
@@ -22,6 +23,15 @@ usage() {
     echo
 }
 
+cleanup_tmp() {
+	if [[ $(pwd) == *vllm-hpu-extension/calibration ]] then
```
Comment (afierka-intel) [inline calibration/calibrate_model.sh:27]: missing semicolon after if statement, before then

Comment (vishnumadhu365) [reply calibration/calibrate_model.sh]: fixed

File: calibration/README.md
Comment (afierka-intel) [inline calibration/README.md:1]: Can you fix the typo as well? Here is missing dash character before `o` parameter:
- `o`, i.e., **output path:** Path to the directory where the generated measurements, etc., will be stored.

File: calibration/README.md [hunk 8]
```diff
@@ -30,3 +30,88 @@ Here are some examples of how to use the script:
...
+# Check if the cluster has required number of HPU's
+ray status
+```
+
+#### Step 3: Run model calibration script
+```bash
+./calibrate_model.sh -m meta-llama/Llama-3.1-405B-Instruct -d <path-to-dataset>/open_orca_gpt4_tokenized_llama.calibration_1000.pkl -o <path-to-calibration-output>/fp8_output -l 4096 -t 16 -b 128
+```
+Running the above command should create the calibration measurement files in the specified output directory with model specific sub-directories.<br>
+<details><summary>Arguments used</summary>
```
Comment (afierka-intel) [inline calibration/README.md:77]: Formatting inconsistency. Can you use formatting as above and use your formatting in first chapter? Just unify this, whatever format you choose, please.

File: calibration/README.md [hunk 9]
```diff
@@ -30,3 +30,88 @@ Here are some examples of how to use the script:
...
+-m for model-id/path<br> 
+-d dataset pickle path<br> 
+-o output directory on nfs<br> 
+-l limit number of data samples used for calibration to the specified value<br> 
+-t tensor parallelism<br> 
+-b batch_size for calibration<br> </details>
+
+<details><summary>Common issues</summary>
```
Comment (afierka-intel) [inline calibration/README.md:85]: The same comment as above. Formatting inconsistency. Can you use formatting as in the frist chapter with TIP clausule, or apply your solution there? Please just decide on one solution and apply it all over the document.

File: calibration/README.md (hunk 8 above)
Comment (vishnumadhu365) [reply calibration/README.md:77]: removed arguments section since its already defined earlier

File: calibration/README.md (hunk 9 above)
Comment (vishnumadhu365) [reply calibration/README.md:85]: removed section, moved it to prerequisites

File: calibration/README.md
Comment (vishnumadhu365) [reply calibration/README.md:1]: fixed

File: calibration/README.md [hunk 10]
```diff
@@ -30,3 +30,76 @@ Here are some examples of how to use the script:
...
+
+#### Step 3: Run model calibration script
+```bash
+./calibrate_model.sh -m meta-llama/Llama-3.1-405B-Instruct -d <path-to-dataset>/open_orca_gpt4_tokenized_llama.calibration_1000.pkl -o <path-to-calibration-output>/fp8_output -l 4096 -t 16 -b 128
+```
+Running the above command should create the calibration measurement files in the specified output directory with model specific sub-directories.<br>
+
+
+#### Step 4: (optional) Measurement unification <p>
+This is an optional step and is used to reduce the target tensor parallelism level by unifying the measurement scales.<br> For eg: You can perform FP8 calibration on the Llama 3.1 405B model on 2x Gaudi2 nodes with Tensor Parallelism = 16 and then use the unification script to reduce the TP to 8. Refer sample command below
```
Comment (michalkuligowski) [inline calibration/README.md:91]: "For eg" Shouldn't it be "e.g." ?

File: calibration/README.md [hunk 11]
```diff
@@ -30,3 +30,76 @@ Here are some examples of how to use the script:
...
+-  `-g` - card grouping to use during unification, card indices separated by commas and groups separated by double dash<br>
+-  `-m` - calibration output path which has the measurement files <br>
+-  `-o` - output directory where unification output gets written<br>
+
+#### Step 5: Serving the FP8 quantized model <p>
+```bash
+export QUANT_CONFIG='<path-to-calibration-output>/fp8_output/llama-3.1-405b-instruct/maxabs_quant_g2.json'
+vllm serve meta-llama/Llama-3.1-405B-Instruct --quantization inc --kv-cache-dtype fp8_inc --weights-load-device cpu --tensor-parallel-size 8
+```
+Note : For serving the output after unification, edit the QUANT_CONFIG file to point the 'dump_stats_path' value to the unification output directory
```
Comment (michalkuligowski) [inline calibration/README.md:104]: There are "Note :" 3 times in the Readme now, should they be formatted using [!NOTE]?

File: calibration/step-2-measure-scales.py [hunk 12]
```diff
@@ -44,6 +44,8 @@ def generate_responses(llm, input_batch, args):
     parser.add_argument("--max-dataset-samples", type=int, default=0)
     parser.add_argument("--max-model-len", type=int, default=2048)
     parser.add_argument("-v", "--verbose", action="store_true")
+    parser.add_argument("--distributed-executor-backend", type=str, default="mp",
```
Comment (michalkuligowski) [inline calibration/step-2-measure-scales.py:47]: Please add "choices=['mp', 'ray']"

File: calibration/step-4-quantize-scales.py [hunk 13]
```diff
@@ -14,6 +14,8 @@
     parser = argparse.ArgumentParser()
     parser.add_argument("--model", type=str, required=True)
     parser.add_argument("--tensor-parallel-size", type=int, default=1)
+    parser.add_argument("--distributed-executor-backend", type=str, default="mp",
```
Comment (michalkuligowski) [inline calibration/step-4-quantize-scales.py:17]: Please add "choices=['mp', 'ray']"

**Comment by vishnumadhu365** [comment]
@michalkuligowski Michal, this PR was merged through https://github.com/HabanaAI/vllm-hpu-extension/pull/100
@afierka-intel Artur, please include these changes in v1.20.0 . Shall I close this PR #92 ?

**Review by michalkuligowski — COMMENTED** [review]
Merged in 100, 1.20 is closed, closing
